tensorcircuit/mpscircuit.py

```diff
@@ -225,35 +298,224 @@ def apply_double_gate(
+        tensor_left = self._mps.tensors[index_left]
+        tensor_right = self._mps.tensors[index_right]
+        new_tensor_left, new_tensor_right = self.reduce_tensor_dimension(
+            tensor_left, tensor_right, center_left=center_left, **self.rules
+        )
+        self._mps.tensors[index_left] = new_tensor_left
+        self._mps.tensors[index_right] = new_tensor_right
+        if center_left:
+            self._mps.center_position = index_left
+        else:
+            self._mps.center_position = index_right
+
+    def apply_MPO(
+        self,
+        tensors: Sequence[Tensor],
+        index_left: int,
+        center_left: bool = True,
+    ) -> None:
+        """
+        Apply a MPO to the MPS
+        """
+        # step 1:
+        #     contract tensor
+        #           a
+        #           |
+        #     i-----O-----j            a
+        #           |        ->        |
+        #           b             ik---X---jl
+        #           |
+        #     k-----T-----l
+        # step 2:
+        #     canonicalize the tensors one by one from end1 to end2
+        # setp 3:
+        #     reduce the bond dimension one by one from end2 to end1
+        nindex = len(tensors)
+        index_right = index_left + nindex - 1
+        if center_left:
+            end1 = index_left
+            end2 = index_right
+            step = 1
+        else:
+            end1 = index_right
+            end2 = index_left
+            step = -1
+        idx_list = np.arange(index_left, index_right + 1)[::step]
+        i_list = np.arange(nindex)[::step]
+
+        self.position(end1)
+        # contract
+        for i, idx in zip(i_list, idx_list):
+            O = tensors[i]
+            T = self._mps.tensors[idx]
+            ni, _, _, nj = O.shape
+            nk, _, nl = T.shape
+            OT = backend.einsum("iabj,kbl->ikajl", O, T)
+            OT = backend.reshape(OT, (ni * nk, 2, nj * nl))
+            self._mps.tensors[idx] = OT
+
+        # canonicalize
+        # FiniteMPS.position applies QR sequentially from index_left to index_right
+        self.position(end2)
+
+        # reduce bond dimension
+        for i in idx_list[::-1][:-1]:
+            self.reduce_dimension(i, i - step, center_left=center_left)
+        assert self._mps.center_position == end1
+
+    def apply_nqubit_gate(
+        self,
+        gate: Gate,
+        *index: int,
+    ) -> None:
+        """
+        Apply a n-qubit gate by transforming the gate to MPO
+        """
+        MPO, index_left = self.gate_to_MPO(gate, *index)
+        index_right = index_left + len(MPO) - 1
+        # start the MPO from the side that the current center is closer to
+        diff_left = abs(index_left - self._mps.center_position)  # type: ignore
+        diff_right = abs(index_right - self._mps.center_position)  # type: ignore
+        self.apply_MPO(MPO, index_left, center_left=diff_left < diff_right)
 
     def apply_general_gate(
```

Comment:
we can support MPO gate in general case, as it is just apply MPO on MPS which we already implemented?

Reply:
I think so

Reply:
pls add a TODO for support of MPO gates

Reply:
What do you mean exactly by MPO gates? the apply_MPO should already do everything needed. Do you mean a API for MPO construction or something like this?

Reply:
the support of such API ``c.mpo(1, 2, mpo=QuOperator)``

Reply:
this is doable via mpo attr in apply_general_gate

Reply:
Again go back to the question of quvector and quoperator... need some discussion